File: itsm-backend/ent/schema/approvalchain.go
package schema

import (
	"time"

	"entgo.io/ent"
	"entgo.io/ent/schema/field"
	"entgo.io/ent/schema/index"
)

// ApprovalChain holds the schema definition for the ApprovalChain entity.
type ApprovalChain struct {
	ent.Schema
}

// Fields of the ApprovalChain.
func (ApprovalChain) Fields() []ent.Field {
	return []ent.Field{
		field.String("name").NotEmpty(),
		field.String("description").Optional(),
		field.String("entity_type").Default("ticket"), // ticket, change, service_request
		field.JSON("chain", []ApprovalChainStep{}),
		field.String("status").Default("active"), // active, inactive
		field.Int("created_by").Optional(),
		field.Int("tenant_id").Optional(),
		field.Time("created_at").Default(time.Now),
		field.Time("updated_at").Default(time.Now).UpdateDefault(time.Now),
	}
}

// ApprovalChainStep 审批链步骤
type ApprovalChainStep struct {
	Level        int    `json:"level"`
	ApproverID   int    `json:"approver_id,omitempty"`
	Role         string `json:"role"`                    // manager, it_admin, security_admin, 或 group:xxx / dept_manager:ID / amount_based:...
	Name         string `json:"name"`                    // 步骤名称
	IsRequired   bool   `json:"is_required"`             // 是否必需
	ApprovalType string `json:"approval_type,omitempty"` // serial: 顺序审批, parallel/all: 会签(AND), or: 或签(OR,默认阈值1)
	Threshold    int    `json:"threshold,omitempty"`     // 会签投票阈值（默认等于审批人数）
	// 兜底策略：当本层解析不到任何审批人且 IsRequired=true 时触发。
	// block(默认,失败关闭): 阻塞待人工介入; auto_approve: 自动通过本层;
	// escalate: 升级到 FallbackApproverID / FallbackRole; auto_reject: 自动拒绝。
	FallbackAction     string `json:"fallback_action,omitempty"`
	FallbackApproverID int    `json:"fallback_approver_id,omitempty"`
	FallbackRole       string `json:"fallback_role,omitempty"`
}

// Edges of the ApprovalChain.
func (ApprovalChain) Edges() []ent.Edge {
	return nil
}

// Indexes of the ApprovalChain.
func (ApprovalChain) Indexes() []ent.Index {
	return []ent.Index{
		index.Fields("tenant_id", "entity_type"),
		index.Fields("status"),
	}
}

File: itsm-frontend/src/types/approval-chain.ts
/**
 * 审批链类型定义
 */

import type { BaseEntity } from './common';

export interface ApprovalChain extends BaseEntity {
  name: string;
  description?: string;
  entityType?: string;
  status?: string;
  isActive: boolean;
  tenantId: number;
  steps: ApprovalStep[];
  createdAt: string;
  updatedAt: string;
}

export interface ApprovalStep extends BaseEntity {
  chainId: number;
  /** 审批层级：同层多个审批人共享同一 level，由 approvalType 决定或签/会签语义 */
  level: number;
  stepOrder: number;
  stepName: string;
  approverType: 'user' | 'role';
  approverId: number;
  approverName: string;
  isRequired: boolean;
  /** 或签 serial：任一人通过即可；会签 parallel：需 threshold 人通过 */
  approvalType?: ApprovalType;
  /** 会签通过人数；为空时后端按该层审批人总数处理（全员通过） */
  threshold?: number;
  /** 无审批人且本层必需时的兜底策略 */
  fallbackAction?: FallbackAction;
  fallbackApproverId?: number;
  fallbackRole?: string;
  /** 动态适配条件：优先级白名单（任一匹配即适用本层） */
  conditionPriorities?: string[];
  /** 适用金额闭区间，0 表示该侧无限制 */
  conditionAmountMin?: number;
  conditionAmountMax?: number;
}

export type ApprovalType = 'serial' | 'parallel';

export type FallbackAction = 'block' | 'auto_approve' | 'escalate' | 'auto_reject';

export const APPROVAL_TYPE_LABELS: Record<ApprovalType, string> = {
  serial: '或签',
  parallel: '会签',
};

export const FALLBACK_ACTION_LABELS: Record<FallbackAction, string> = {
  block: '阻断（无人可审时报错）',
  escalate: '升级到指定审批人/角色',
  auto_approve: '自动通过',
  auto_reject: '自动拒绝',
};

export const APPROVAL_CONDITION_PRIORITIES = ['low', 'medium', 'high', 'urgent', 'critical'] as const;

export interface ApprovalChainStats {
  total: number;
  active: number;
  inactive: number;
  totalSteps: number;
  avgStepsPerChain: number;
}

/** 只承载后端真实支持的筛选条件：name 为子串匹配，entityType/status 为精确匹配 */
export interface ApprovalChainFilters {
  name?: string;
  entityType?: string;
  status?: 'active' | 'inactive';
}

export interface CreateApprovalChainRequest {
  name: string;
  description?: string;
  entityType?: string;
  isActive: boolean;
  steps: Omit<ApprovalStep, 'id' | 'chainId' | 'createdAt' | 'updatedAt'>[];
}

export interface UpdateApprovalChainRequest {
  name?: string;
  description?: string;
  entityType?: string;
  isActive?: boolean;
  steps?: Omit<ApprovalStep, 'id' | 'chainId' | 'createdAt' | 'updatedAt'>[];
}
